### ISSUE TITLE
Slow Query: Notification Read Inserts by User

### ISSUE THREAD

**Issue by NakulManchanda** [hicommonwealth/commonwealth#3438]
## Description

The way we do read notification - leads to very slow query, which is very frequently used

 "Notifications Read" table is huge ~21M rows in
`SELECT count(1) from "NotificationsRead";` this query itself takes 20sec

https://github.com/hicommonwealth/commonwealth/blob/6faa3108006615380ad05193e91578052586bbf5/packages/commonwealth/server/util/emitNotifications.ts#L165

Need to find better way to keep max notification read

```SQL
INSERT INTO "NotificationsRead" 
VALUES ($1, $2, $3, $4, (SELECT COALESCE(MAX(id), $5) + $6 FROM "NotificationsRead" WHERE user_id = $7))
, ($8, $9, $10, $11, (SELECT COALESCE(MAX(id), $12) + $13 FROM "NotificationsRead" WHERE user_id = $14))
, ($15, $16, $17, $18, (SELECT COALESCE(MAX(id), $19) + $20 FROM "NotificationsRead" WHERE user_id = $21))
, ($22, $23, $24, $25, (SELECT COALESCE(MAX(id), $26) + $27 FROM "NotificationsRead" WHERE user_id = $28))
, ($29, $30, $31, $32, (SELECT COALESCE(MAX(id), $33) + $34 FROM "NotificationsRead" WHERE user_id = $35))
, ($36, $37, $38, $39, (SELECT COALESCE(MAX(id), $40) + $41 FROM "NotificationsRead" WHERE user_id = $42))
, ($43, $44, $45, $46, (SELECT COALESCE(MAX(id), $47) + $48 FROM "NotificationsRead" WHERE user_id = $49))
, ($8933, $8934, $8935, $8936, (SELECT COALESCE(MAX(id), $8937) + $8938 FROM "NotificationsRead" WHERE user_id = $8939))
, ($8940, $8941, $8942, $8943, (SELECT COALESCE(MAX(id), $8944) + $8945 FROM "NotificationsRead" WHERE user_id = $8946))
, ($8947, $8948, $8949, $8950, (SELECT COALESCE(MAX(id), $8951) + $8952 FROM "NotificationsRead" WHERE user_id = $8953))
```

query happens in bulk
- SELECT COALESCE(MAX(id), $8951) + $8952 FROM "NotificationsRead" WHERE user_id = $8953

## Acceptance Criteria
- Fast Read Notification Receipts

## Possible Solution(s)
- Find different way to keep read receipts
- Partition Table - to find Max in faster way
- or keep max pre-computed
- another pattern which could be used
- make it a two step query
   - Find all Max with max_ids as `SELECT user_id, max(id) as max_id FROM NotificationRead GROUP BY user_id`
   - then insert `INSERT INTO "NotificationsRead" SELECT ?,?,?,?,max_id+1 from max_ids JOIN subscriptions` etc

**Comment by NakulManchanda** [comment]
Thinking of reworking schema for this.
- keeping track of max by user_id by subscription_id

New schema might be able to improve performance for 
`viewUserActivity` , `viewGlobalActivity`, as well as this slow update

**Comment by jnaviask** [comment]
Given that the change from NotificationsRead to max notifications would have product impact, I'm tagging @zakhap on this ticket to discuss whether the change to id is feasible.

Would love to get a % estimation on the performance gain for the query in a max-based system.

**Comment by zakhap** [comment]
The tradeoff of losing notification-level granular "mark as read" needs to be justified by a measurable performance boost (which this seems to be, undeniably). A rough % estimate here would be greatly helpful in making the decision here.

**Comment by NakulManchanda** [comment]
We are keeping all tables:
- adding extra table or column in some existing table to keep track of max

New Table - NotificationReadMax
- userId, thread_id, max_not_id

NotificationRead
- stay same

Current Query: Read max on every row insert for every user
```
- (SELECT (SELECT MAX(by <user_id>) )+1 
(SELECT MAX(by <user_id>) )+1 
```

New Query:
Transaction:
   1. Get max for all users SELECT NotificationReadMax FOR UPDATE // this will lock rows for updates
      ``` 
      SELECT FORM NotificationReadMax where EXISTS(user_id=<> and thread_id=<thread_id>)
      ```
   2. inserts all new rows in  NotificationRead 
   3. Update  new Max by user_id & thread_id in NotificationReadMax
   4. Release Transaction
   
 Additional Benefit if turns out way we suppose it can go:
 `viewUserActivity` `viewGlobalActivity`
 - we can get rid of calculating max & groupby - which is slower part of both queries
 
 ```
WITH ranked_thread_notifs as (SELECT n.thread_id                                 AS thread_id,
                                           MAX(n.id)                                   as mx_not_id,
                                           ROW_NUMBER() OVER (ORDER BY MAX(n.id) DESC) as thread_rank
                                    FROM "Notifications" n
                                    WHERE n.category_id IN ('new-thread-creation', 'new-comment-creation')
                                      AND n.chain_id IN (SELECT a."chain" FROM "Addresses" a WHERE a.user_id = ?)
                                      AND n.thread_id IS NOT NULL
                                    GROUP BY n.thread_id)
```

new 
```
FROM "Notifications" n
JOIN NotificationReadMax nm on n.user_id=nm.user_id and n.thread_id=nm.thread_id
                                    WHERE n.category_id IN ('new-thread-creation', 'new-comment-creation')
                                      AND n.chain_id IN (SELECT a."chain" FROM "Addresses" a WHERE a.user_id = ?)
                                      AND n.thread_id IS NOT NULL

```

We just want to improve Reads:
- we can keep track of max in After Insert Trigger but that might make update even slower, read could be faster though


Basically just keeping track of MAX properly to even rely on it for updates:
New Table - NotificationReadMax
- userId, thread_id, max_not_id
 
**If we only want to target reads,** there is another approach:
- **Materialized View** -> Just refresh materialized view after all updates are done - its more like caching but at database level
- values might be little stale, if we can tolerate some staleness -> as view would be refresh only on updates
- we might run select query for every user every thread atleast once after every update 
- "Incremental Materailized View" (need to check support) can help with faster updates - but query will still run for all user all threads all user after every NotificationRead Update

**Comment by jnaviask** [comment]
@NakulManchanda this doesn't answer the question of a % performance gain in estimation

**Comment by NakulManchanda** [comment]
50% improvement atleast is expected out of any change is what i aim for
- both read & update

- Improvement in both update(reduce load on whole notification side)
   Update Performance is degrading too with more users, threads & notification data
- Reads will have faster way too!
   Reads would be slower overtime too without change
   

We have to do a POC.
I can try out a quick POC by making suggested schema changes manually to measure performance!

**Comment by NakulManchanda** [comment]
This improvement is simillar to running for loop on SELECT calls
In this query we are running for loop on UPDATE calls - that too **group by** max on every insert!

Denormalize data by keeping track of max in separate table is the core idea.

**Comment by zakhap** [comment]
> I can try out a quick POC by making suggested schema changes manually to measure performance!

This would be hugely helpful in making the product decision around the change here.

**Comment by jnaviask** [comment]
Feel free to proceed with a POC.

**Comment by zakhap** [comment]
@NakulManchanda When can I expect a POC for the improvement here?

**Comment by NakulManchanda** [comment]
This week/early next - i will keep you posted

**Comment by zakhap** [comment]
related work: #4164

**Comment by NakulManchanda** [comment]
Ok, Thanks @timolegros - this issue attached to new PR now.
